src/modules/iotjs_module_websocket.c

```diff
@@ -0,0 +1,737 @@
+/* Copyright 2018-present Samsung Electronics Co., Ltd. and other contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <time.h>
+
+#include "iotjs_def.h"
+#include "iotjs_handlewrap.h"
+#include "iotjs_module_buffer.h"
+#include "iotjs_module_websocket.h"
+#include "iotjs_reqwrap.h"
+#include "mbedtls/base64.h"
+#include "mbedtls/sha1.h"
+
+
+static void iotjs_wsclient_destroy(iotjs_wsclient_t *wsclient) {
+  IOTJS_RELEASE(wsclient->tcp_buff.buffer);
+  IOTJS_RELEASE(wsclient->ws_buff.data);
+  IOTJS_RELEASE(wsclient);
+}
+
+
+static const jerry_object_native_info_t wsclient_native_info = {
+  .free_cb = (jerry_object_native_free_callback_t)iotjs_wsclient_destroy
+};
+
+
+iotjs_wsclient_t *iotjs_wsclient_create(const jerry_value_t jobject) {
+  iotjs_wsclient_t *wsclient = IOTJS_ALLOC(iotjs_wsclient_t);
+
+  jerry_set_object_native_pointer(jobject, wsclient, &wsclient_native_info);
+  return wsclient;
+}
+
+
+static const char WS_GUID[] = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+static unsigned char *generated_key = NULL;
+
+/**
+ * The protocol is as follows:
+ * method + USER_ENDPOINT + protocol
+ * host + USER_HOST + line_end
+ * upgrade
+ * connection
+ * sec_websocket_key + line_end
+ * sec_websocket_ver
+ */
+static const char method[] = "GET ";
+static const char protocol[] = " HTTP/1.1\r\n";
+static const char host[] = "Host: ";
+static const char line_end[] = "\r\n";
+static const char upgrade[] = "Upgrade: websocket\r\n";
+static const char connection[] = "Connection: Upgrade\r\n";
+static const char sec_websocket_key[] = "Sec-WebSocket-Key: ";
+static const char sec_websocket_ver[] = "Sec-WebSocket-Version: 13\r\n\r\n";
+static size_t header_fixed_size =
+    sizeof(method) + sizeof(protocol) + sizeof(host) + sizeof(upgrade) +
+    sizeof(connection) + sizeof(sec_websocket_key) + sizeof(sec_websocket_ver) -
+    9; // 9 is for every \0
+
+
+static void iotjs_websocket_create_callback(jerry_value_t jsref,
+                                            jerry_value_t jmsg, char *name) {
+  iotjs_jargs_t args;
+  args = iotjs_jargs_create(1);
+  if (!jerry_value_is_undefined(jmsg)) {
+    iotjs_jargs_append_jval(&args, jmsg);
+  } else {
+    iotjs_jargs_append_jval(&args, jerry_create_undefined());
+  }
+
+  jerry_value_t fn = iotjs_jval_get_property(jsref, name);
+  iotjs_make_callback(fn, jsref, &args);
```

Comment:
We are started to remove the `iotjs_jargs_t`. Please use `iotjs_invoke_callback` here.